tests/proxy-non-sharable-responses.js

```diff
@@ -0,0 +1,136 @@
+        testCase.client().request.for(resource);
+        testCase.server().serve(resource);
+        testCase.server().response.header.add("Cache-Control", "max-age=0");
+        await testCase.run();
+    }
+    
+    async doSingleCheck(revalidation, makePrivate)
+    {
+        const clientsCount = Number.parseInt(Config.ClientsCount, 10);
+    
+        let resource = new Resource();
+        resource.uri.address = AddressPool.ReserveListeningAddress();
+        resource.modifiedAt(FuzzyTime.DistantPast());
+        resource.finalize();
+    
+        if (revalidation)
+            await this.cacheSomething(resource);
+    
+        let testCase = new HttpTestCase("send " + clientsCount.toString() + " requests to make Squid collapse on them");
+        testCase.client().request.for(resource);
+        testCase.makeClients(clientsCount - 1, (client => {
+            client.request.for(resource);
+        }));
+    
+        testCase.server().serve(resource);
+        if (makePrivate)
+            makePrivate(testCase, resource);
+        testCase.server().transaction().blockSendingUntil(
+                testCase.clientsSentEverything(),
+                "wait for all clients to collapse");
+
+        testCase.check(() => {
+            const clients = testCase.clients();
+            const parentID = clients[0].transaction().request.id();
+            for (let i = 0; i < clients.length; ++i) {
+                let request = clients[i].transaction().request;
+                let response = clients[i].transaction().response;
+                const sentID = request.id();
+                const receivedID = response.extractMatchingId(request);
+                const statusCode = response.startLine.codeString();
+                let msg = "changed X-Daft-Response-Tag, private headers ";
+                msg += makePrivate ? "on" : "off";
+                if (i === 0) {
+                    assert.equal(statusCode, 200);
+                    assert.equal(parentID, receivedID, msg);
+                } else {
+                    if (!makePrivate)
+                        assert.equal(parentID, receivedID, msg);
+                    else // expecting 200 or 50x
+                        assert(parentID !== receivedID);
+                }
+            }
+        });
+
+        await testCase.run();
+    
+        AddressPool.ReleaseListeningAddress(resource.uri.address);
+    }
+
+    async doCheck(revalidation, isPrivate) {
+        await this.doSingleCheck(revalidation, isPrivate ? this.authClt : null);
+        await this.doSingleCheck(revalidation, isPrivate ? this.noStoreClt : null);
+        await this.doSingleCheck(revalidation, isPrivate ? this.noStoreSrv : null);
+        await this.doSingleCheck(revalidation, isPrivate ? this.privateSrv : null);
+    }
+
+    async run(/*testRun*/) {
+        console.log("Test A: the proxy must support entries sharing(collapsing)");
+        await this.doCheck(false, false);
```

Comment:
The other test covers all that Test A does. IIRC we decided (long ago) that it would be useful to ran the same test with positive outcome first thus making this test self-sufficient, testing both positive and negative outcomes (since it is relatively easy to do).